db/db_impl/db_impl_write.cc

```diff
@@ -60,6 +66,145 @@ Status DBImpl::WriteWithCallback(const WriteOptions& write_options,
+    }
+    autovector<WriteBatch*> batches;
+    for (auto w : updates) {
+      batches.push_back(w);
+    }
+    return MultiThreadWriteImpl(options, batches, nullptr, nullptr);
+  } else {
+    Status status;
+    for (auto b : updates) {
+      status = WriteImpl(options, b, nullptr, nullptr);
+      if (!status.ok()) {
+        return status;
+      }
+    }
+    return status;
+  }
+}
+
+Status DBImpl::MultiThreadWriteImpl(const WriteOptions& write_options,
+                                    const autovector<WriteBatch*>& updates,
+                                    WriteCallback* callback, uint64_t* log_used,
+                                    uint64_t log_ref, uint64_t* seq_used) {
+  StopWatch write_sw(env_, immutable_db_options_.statistics.get(), DB_WRITE);
+  WriteThread::Writer writer(write_options, updates, callback, log_ref);
+  write_thread_.JoinBatchGroup(&writer);
+
+  WriteContext write_context;
+  PERF_TIMER_GUARD(write_pre_and_post_process_time);
+  bool ignore_missing_faimly = write_options.ignore_missing_column_families;
+  if (writer.state == WriteThread::STATE_GROUP_LEADER) {
+    if (writer.callback && !writer.callback->AllowWriteBatching()) {
+      write_thread_.WaitForMemTableWriters();
+    }
+    WriteThread::WriteGroup wal_write_group;
+    mutex_.Lock();
+    bool need_log_sync = !write_options.disableWAL && write_options.sync;
+    bool need_log_dir_sync = need_log_sync && !log_dir_synced_;
+    PERF_TIMER_STOP(write_pre_and_post_process_time);
+    writer.status =
+        PreprocessWrite(write_options, &need_log_sync, &write_context);
+    PERF_TIMER_START(write_pre_and_post_process_time);
+    log::Writer* log_writer = logs_.back().writer;
+    mutex_.Unlock();
+
+    // This can set non-OK status if callback fail.
+    last_batch_group_size_ =
+        write_thread_.EnterAsBatchGroupLeader(&writer, &wal_write_group);
+    const SequenceNumber current_sequence =
+        write_thread_.UpdateLastSequence(versions_->LastSequence()) + 1;
+    size_t total_count = 0;
+    size_t total_byte_size = 0;
+    size_t valid_batches = 0;
+    auto stats = default_cf_internal_stats_;
+    if (writer.status.ok()) {
+      SequenceNumber next_sequence = current_sequence;
+      for (auto w : wal_write_group) {
+        if (w->CheckCallback(this)) {
+          if (w->ShouldWriteToMemtable()) {
+            w->sequence = next_sequence;
+            size_t count = WriteBatchInternal::Count(w->batches);
+            next_sequence += count;
+            total_count += count;
+          }
+          valid_batches += 1;
+          total_byte_size = WriteBatchInternal::AppendedByteSize(
+              total_byte_size, WriteBatchInternal::ByteSize(w->batches));
+        }
+      }
+      if (writer.disable_wal) {
+        has_unpersisted_data_.store(true, std::memory_order_relaxed);
+      }
+      write_thread_.UpdateLastSequence(current_sequence + total_count - 1);
+      if (!write_options.disableWAL) {
```

Comment:
These stats seems missing? https://github.com/facebook/rocksdb/blob/2afea29762a214f099e1e4749b04839646657478/db/db_impl/db_impl_write.cc#L509

Reply:
kIntStatsNumKeysWritten? I will fix it.

Reply:
yup